mingw32 build fails due to missing nix::readLine()

**Describe the bug**

https://hydra.nixos.org/build/277363338

```
[6/6] Linking target libnixmain.dll
FAILED: libnixmain.dll 
x86_64-w64-mingw32-g++  -o libnixmain.dll libnixmain.dll.p/common-args.cc.obj libnixmain.dll.p/loggers.cc.obj libnixmain.dll.p/plugin.cc.obj libnixmain.dll.p/progress-bar.cc.obj libnixmain.dll.p/shared.cc.obj -Wl,--allow-shlib-undefined -shared -Wl,--start-group -Wl,--out-implib=libnixmain.dll.a /nix/store/jdqqx9nnz5yp0myjyjrblbdk8hxk60kc-nix-util-x86_64-w64-mingw32-2.25.0pre/lib/libnixutil.dll.a /nix/store/b2n32ylfcw3dsxznj81akd8539ad4vlb-libarchive-x86_64-w64-mingw32-3.7.4-lib/lib/libarchive.dll.a /nix/store/llhkczan46l8x9v11sv7cb36p6nrkhhi-nix-store-x86_64-w64-mingw32-2.25.0pre/lib/libnixstore.dll.a -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -Wl,--end-group
/nix/store/4bkj4cch6qjy02254b3dwpgc7sfkk9v0-x86_64-w64-mingw32-binutils-2.41/bin/x86_64-w64-mingw32-ld: libnixmain.dll.p/progress-bar.cc.obj:progress-bar.c:(.text$_ZN3nix11ProgressBar3askESt17basic_string_viewIcSt11char_traitsIcEE[_ZN3nix11ProgressBar3askESt17basic_string_viewIcSt11char_traitsIcEE]+0x103): undefined reference to `nix::readLine[abi:cxx11](void*, bool)'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

**Steps To Reproduce**

1. Go to '...'
2. Click on '....'
3. Scroll down to '....'
4. See error

**Expected behavior**

A clear and concise description of what you expected to happen.

**`nix-env --version` output**

**Additional context**

Add any other context about the problem here.

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

--- a/src/libmain/progress-bar.cc
+++ b/src/libmain/progress-bar.cc
@@ -543,7 +543,7 @@ public:
         auto state(state_.lock());
         if (!state->active) return {};
         std::cerr << fmt("\r\e[K%s ", msg);
-        auto s = trim(readLine(STDIN_FILENO, true));
+        auto s = trim(readLine(getStandardInput(), true));
         if (s.size() != 1) return {};
         draw(*state);
         return s[0];

--- a/src/libutil/file-descriptor.hh
+++ b/src/libutil/file-descriptor.hh
@@ -106,12 +106,42 @@ void drainFD(
 #endif
     );
 
+/**
+ * Get [Standard Input](https://en.wikipedia.org/wiki/Standard_streams#Standard_input_(stdin))
+ */
 [[gnu::always_inline]]
-inline Descriptor getStandardOut() {
+inline Descriptor getStandardInput()
+{
+#ifndef _WIN32
+    return STDIN_FILENO;
+#else
+    return GetStdHandle(STD_INPUT_HANDLE);
+#endif
+}
+
+/**
+ * Get [Standard Output](https://en.wikipedia.org/wiki/Standard_streams#Standard_output_(stdout))
+ */
+[[gnu::always_inline]]
+inline Descriptor getStandardOutput()
+{
 #ifndef _WIN32
     return STDOUT_FILENO;
 #else
     return GetStdHandle(STD_OUTPUT_HANDLE);
+#endif
+}
+
+/**
+ * Get [Standard Error](https://en.wikipedia.org/wiki/Standard_streams#Standard_error_(stderr))
+ */
+[[gnu::always_inline]]
+inline Descriptor getStandardError()
+{
+#ifndef _WIN32
+    return STDERR_FILENO;
+#else
+    return GetStdHandle(STD_ERROR_HANDLE);
 #endif
 }
 

--- a/src/libutil/logging.cc
+++ b/src/libutil/logging.cc
@@ -38,7 +38,7 @@ void Logger::warn(const std::string & msg)
 
 void Logger::writeToStdout(std::string_view s)
 {
-    Descriptor standard_out = getStandardOut();
+    Descriptor standard_out = getStandardOutput();
     writeFull(standard_out, s);
     writeFull(standard_out, "\n");
 }
@@ -118,11 +118,7 @@ void writeToStderr(std::string_view s)
 {
     try {
         writeFull(
-#ifdef _WIN32
-            GetStdHandle(STD_ERROR_HANDLE),
-#else
-            STDERR_FILENO,
-#endif
+            getStandardError(),
             s, false);
     } catch (SystemError & e) {
         /* Ignore failing writes to stderr.  We need to ignore write

--- a/src/libutil/windows/file-descriptor.cc
+++ b/src/libutil/windows/file-descriptor.cc
@@ -61,7 +61,7 @@ void writeFull(HANDLE handle, std::string_view s, bool allowInterrupts)
 }
 
 
-std::string readLine(HANDLE handle)
+std::string readLine(HANDLE handle, bool eofOk)
 {
     std::string s;
     while (1) {
@@ -71,8 +71,12 @@ std::string readLine(HANDLE handle)
         DWORD rd;
         if (!ReadFile(handle, &ch, 1, &rd, NULL)) {
             throw WinError("reading a line");
-        } else if (rd == 0)
-            throw EndOfFile("unexpected EOF reading a line");
+        } else if (rd == 0) {
+            if (eofOk)
+                return s;
+            else
+                throw EndOfFile("unexpected EOF reading a line");
+        }
         else {
             if (ch == '\n') return s;
             s += ch;

--- a/src/nix-store/nix-store.cc
+++ b/src/nix-store/nix-store.cc
@@ -694,7 +694,7 @@ static void opDump(Strings opFlags, Strings opArgs)
     if (!opFlags.empty()) throw UsageError("unknown flag");
     if (opArgs.size() != 1) throw UsageError("only one argument allowed");
 
-    FdSink sink(getStandardOut());
+    FdSink sink(getStandardOutput());
     std::string path = *opArgs.begin();
     dumpPath(path, sink);
     sink.flush();
@@ -722,7 +722,7 @@ static void opExport(Strings opFlags, Strings opArgs)
     for (auto & i : opArgs)
         paths.insert(store->followLinksToStorePath(i));
 
-    FdSink sink(getStandardOut());
+    FdSink sink(getStandardOutput());
     store->exportPaths(paths, sink);
     sink.flush();
 }
@@ -835,7 +835,7 @@ static void opServe(Strings opFlags, Strings opArgs)
     if (!opArgs.empty()) throw UsageError("no arguments expected");
 
     FdSource in(STDIN_FILENO);
-    FdSink out(getStandardOut());
+    FdSink out(getStandardOutput());
 
     /* Exchange the greeting. */
     ServeProto::Version clientVersion =

--- a/src/nix/cat.cc
+++ b/src/nix/cat.cc
@@ -16,7 +16,7 @@ struct MixCat : virtual Args
             throw Error("path '%1%' is not a regular file", path);
         stopProgressBar();
 
-        writeFull(getStandardOut(), accessor->readFile(CanonPath(path)));
+        writeFull(getStandardOutput(), accessor->readFile(CanonPath(path)));
     }
 };
 

--- a/src/nix/dump-path.cc
+++ b/src/nix/dump-path.cc
@@ -20,7 +20,7 @@ struct CmdDumpPath : StorePathCommand
 
     void run(ref<Store> store, const StorePath & storePath) override
     {
-        FdSink sink(getStandardOut());
+        FdSink sink(getStandardOutput());
         store->narFromPath(storePath, sink);
         sink.flush();
     }
@@ -55,7 +55,7 @@ struct CmdDumpPath2 : Command
 
     void run() override
     {
-        FdSink sink(getStandardOut());
+        FdSink sink(getStandardOutput());
         dumpPath(path, sink);
         sink.flush();
     }

--- a/src/nix/eval.cc
+++ b/src/nix/eval.cc
@@ -115,7 +115,7 @@ struct CmdEval : MixJSON, InstallableValueCommand, MixReadOnlyOption
 
         else if (raw) {
             stopProgressBar();
-            writeFull(getStandardOut(), *state->coerceToString(noPos, *v, context, "while generating the eval command output"));
+            writeFull(getStandardOutput(), *state->coerceToString(noPos, *v, context, "while generating the eval command output"));
         }
 
         else if (json) {

--- a/src/nix/log.cc
+++ b/src/nix/log.cc
@@ -57,7 +57,7 @@ struct CmdLog : InstallableCommand
             if (!log) continue;
             stopProgressBar();
             printInfo("got build log for '%s' from '%s'", installable->what(), logSub.getUri());
-            writeFull(getStandardOut(), *log);
+            writeFull(getStandardOutput(), *log);
             return;
         }
 

--- a/src/nix/sigs.cc
+++ b/src/nix/sigs.cc
@@ -177,7 +177,7 @@ struct CmdKeyGenerateSecret : Command
             throw UsageError("required argument '--key-name' is missing");
 
         stopProgressBar();
-        writeFull(getStandardOut(), SecretKey::generate(*keyName).to_string());
+        writeFull(getStandardOutput(), SecretKey::generate(*keyName).to_string());
     }
 };
 
@@ -199,7 +199,7 @@ struct CmdKeyConvertSecretToPublic : Command
     {
         SecretKey secretKey(drainFD(STDIN_FILENO));
         stopProgressBar();
-        writeFull(getStandardOut(), secretKey.toPublicKey().to_string());
+        writeFull(getStandardOutput(), secretKey.toPublicKey().to_string());
     }
 };
 

